Threescale 2689 service subscription api

nidhi-soni1104 opened the pull request:
<!--  Thanks for sending a pull request!  Here are some tips for you:
1. Please remember to ALWAYS open an issue before starting to work on your pull request. Please take the time to validate your intentions for the pull request with the project maintainers before spending the time to work on it, so your time does not go to waste. 
2. If this is your first time, please make sure you've gone through the Contribution guide.
3. If the PR is unfinished, add a `[WIP]` at the start of the PR title. You can remove it when it's ready to be reviewed.
-->

**What this PR does / why we need it**:

Added in Jira 

**Which issue(s) this PR fixes** 

https://issues.redhat.com/browse/THREESCALE-2689

https://issues.redhat.com/browse/THREESCALE-10728

mayorova commented on doc/active_docs/account_management_api.json:
```diff
@@ -1306,6 +1360,63 @@
+          {
+            "name": "id",
+            "in": "path",
+            "description": "ID of the fields definition.",
```
"fields definition"? :upside_down_face:

nidhi-soni1104 replied:
oh that's incorrect, but updated now

mayorova commented on doc/active_docs/account_management_api.json:
```diff
@@ -1261,6 +1261,60 @@
+            }
+          },
+          {
+            "name": "service_id",
```
1. It is not recommended (and probably just wrong) to have both query and body parameters in a POST request.
2. if we have the plan ID of the service plan, we don't actually need the service ID, because if would be inferred from the service plan.

nidhi-soni1104 replied:
agreed with both points we can remove the service ID

nidhi-soni1104 replied:
This PR need more changes I am working on it

mayorova commented on doc/active_docs/account_management_api.json:
```diff
@@ -1306,6 +1356,67 @@
+                  },
+                  "service_id": {
+                    "type": "integer",
+                    "description": "ID of the service."
```
Missing `}` here

mayorova commented on app/controllers/admin/api/service_contracts_controller.rb:
```diff
@@ -21,14 +29,43 @@ def destroy
+  end
+
+  def service
+    service ||= accessible_services.find(params[:service_id])
```
I mentioned this before, but I'll put it here again - I don't think we need `service_id`, I think it's redundant, because we have a `plan_id`. Another thing is that we'd need to validate this plan ID:
- make sure that it is actually a Service plan
- verify that the user sending the request has permissions on this Service

If these conditions are not met, a proper error must be shown. Currently, if both `service_id` and `plan_id` are set, but there is a mismatch (i.e. the plan doesn't belong to the service specified by the ID), the following error is shown:
```
<?xml version="1.0" encoding="UTF-8"?>
  <errors>
    <error>Plan plan must be a ServicePlan</error>
  </errors>
```

This is confusing, because the plan ID does actually belong to a ServicePlan.

But again, I'd just remove the service ID completely.

nidhi-soni1104 replied:
@mayorova I upadted create action.

mayorova commented on app/controllers/admin/api/service_contracts_controller.rb:
```diff
@@ -21,14 +29,43 @@ def destroy
+  end
+
+  def service_plan(plan_id = service_contract_plan_id)
+    service_plan ||= service.service_plans.find_by(id: plan_id)
```
`service_plan` is a local variable, which is never reused. I guess you meant `@service_plan ||=... ` ?

mayorova reviewed (changes requested):
1. We also need a GET for the specific service contract
2. Tests are required
3. Inline comments

nidhi-soni1104 commented:
Create API is updated and no service ID is used

1.checked condition if wrong account id given
2. checked for valid plans and its account it (this will verify that the user sending the request has permissions on this Service) 

more suggestions appreciated

nidhi-soni1104 commented:
@mayorova Test are added and PR is ready to review
for this **We also need a GET for the specific service contract** I am  creating new Jira and will create new PR for the same.

mayorova commented:
> for this **We also need a GET for the specific service contract** I have created new JIRA and will create new PR for this

I think it needs to be added in this PR, without it the feature is not quite complete (CRUD without the Read), and I think it's just a couple of lines of code, and another couple for tests.

mayorova commented on app/controllers/admin/api/service_contracts_controller.rb:
```diff
@@ -6,29 +6,91 @@ class Admin::Api::ServiceContractsController < Admin::Api::ServiceBaseController
 
-  def account
-    @account ||= current_account.buyers.find params.require(:account_id)
+  def find_account
+    @account ||= current_account.buyers.find_by(id: params.require(:account_id))
+
+    unless @account
+      render_error('Buyer not found with this account ID', status: :not_found)
+    end
```
```suggestion
  def account
    @account ||= current_account.buyers.find params.require(:account_id)
  end
```

I think this change is not needed, the original code was working just fine.
So, `before_action :find_account` is not needed either, and `@account` can be changed to `account`.

`#find` method throws a `ActiveRecord::RecordNotFound`, which is handled [here](https://github.com/3scale/porta/blob/e50efabc23679892d43115a9d2238e864ead7d93/app/controllers/concerns/error_handling.rb#L8), that makes the API request to return 404.

I think we don't need to include the error message (and moreover we should avoid the term "buyer" whenever possible), because all other API endpoints just return 404 with no error message.

nidhi-soni1104 replied:
Sounds good but I added so whoever use API can clearly know okay this input is wrong

mayorova commented on app/controllers/admin/api/service_contracts_controller.rb:
```diff
@@ -6,29 +6,91 @@ class Admin::Api::ServiceContractsController < Admin::Api::ServiceBaseController
 
-  def service_contract
-    @service_contract ||= account.bought_service_contracts.find params.require(:id)
+  def find_service_contract
+    @service_contract ||= @account.bought_service_contracts.find_by(id: params.require(:id))
+
+    unless @service_contract
+      render_error('Service contract not found', status: :not_found)
+    end
```
```suggestion
  def service_contract
    @service_contract ||= account.bought_service_contracts.find params.require(:id)
```

Same as above - we can keep the old code, rename `@service_contract` to `service_contract`, and remove the corresponding `before_action`.

mayorova commented on config/routes.rb:
```diff
@@ -611,7 +611,7 @@
         end
 
-        resources :service_contracts, :only => [:index, :destroy]
+        resources :service_contracts, :only => [:index, :create, :update, :destroy, :show]
```
```suggestion
        resources :service_contracts, only: %i[index create update destroy show]
```

Just better Ruby style

mayorova commented on app/controllers/admin/api/service_contracts_controller.rb:
```diff
@@ -21,14 +34,52 @@ def destroy
+  end
+
+  def create_service_contract(plan)
+    @service_contract = account.bought_service_contracts.create(service_contract_params.merge(plan: plan))
+    respond_with @service_contract
```
Here we don't need to use this as an instance variable `@service_contract`, just a local one, e.g. `new_contract`
In fact:
1. We don't need an extra variable at all, and can just do 
```
respond_with account.bought_service_contracts.create(service_contract_params.merge(plan: plan))
```

2. We don't need this extra method, and can move this line inside the `#create`.

mayorova replied:
And also, `service_contract_params` already merges in the `plan` instance, don't need to merge it in again.

mayorova commented on app/controllers/admin/api/service_contracts_controller.rb:
```diff
@@ -7,6 +7,19 @@ class Admin::Api::ServiceContractsController < Admin::Api::ServiceBaseController
+  def create
+    plan = service_plan
+
+    unless valid_plan_and_account?(plan)
+      render_error('Invalid service plan or account id', status: :unprocessable_entity)
+      return
+    end
```
This can be moved to the `before_action` hook

mayorova commented on app/controllers/admin/api/service_contracts_controller.rb:
```diff
@@ -21,14 +34,52 @@ def destroy
   end
 
+  def service_plan
+    ServicePlan.find_by(id: service_contract_plan_id)
```
This also needs to be `.find` to throw NotFound exception

mayorova commented:
I've added some suggestions here: https://github.com/3scale/porta/pull/3694

Basically, trying to make the code easier to follow, with less duplicate logic (fetching plan by ID in different ways), etc.

mayorova commented on test/integration/user-management-api/service_contracts_controller_test.rb:
```diff
@@ -25,6 +30,44 @@ def test_index
+        account_id: @buyer.id,
+        format: :xml,
+        access_token: @token,
+        service_contract: { plan_id: @new_service_plan.id }
```
I might be wrong, but I think we shouldn't wrap `plan_id` into `service_contract`, instead we need to verify that the controller itself wraps it properly.

nidhi-soni1104 replied:
Your observation is correct. In a well-structured Rails controller, the parameters should be properly handled by the controller itself, and you should not need to explicitly wrap the plan_id into service_contract in the test.

but here its not doing so I have done it explicitily

mayorova commented on test/integration/user-management-api/service_contracts_controller_test.rb:
```diff
@@ -7,10 +7,15 @@ def setup
     @service_contract = FactoryBot.create(:simple_service_contract, plan: @service_plan, user_account: @buyer)
 
     @buyer.buy! @application_plan
+    @buyer.buy! @new_application_plan
```
```suggestion
```

This step is not needed, and in fact it is very confusing. In real life, a buyer would never have an application in a service when they don't have a service subscription.

mayorova commented on test/integration/user-management-api/service_contracts_controller_test.rb:
```diff
@@ -25,6 +30,44 @@ def test_index
+        account_id: @buyer.id,
+        format: :xml,
+        access_token: @token,
+        service_contract: { plan_id: @service_plan.id }
```
I think it would be better to use a different service plan (under the same service)

mayorova commented on test/integration/user-management-api/service_contracts_controller_test.rb:
```diff
@@ -25,6 +30,44 @@ def test_index
+      assert_match "already subscribed to this service", response.body
+    end
+
+    def test_failure_subscribe
```
What is the test about? (i.e. what is the issue)
and where are the assertions? :thinking: 

I think along with the successful flow, we need to test subscription to a service plan that doesn't belong to the buyer's provider (should fail).

nidhi-soni1104 replied:
yes assertion for not found is missing here

this test does the same  you said (we need to test subscription to a service plan that doesn't belong to the buyer's provider (should fail).). 
service_plan doesn't belongs to current_account but belongs to buyer and so fails 
I have added assertion and comment also I updated name to make it more clear

mayorova reviewed (changes requested):
Some suggestions inline, plus the following:

1. There are no tests for "update subscription" scenario. We need to validate both success and failure flows.

2. I think we need to review the member permissions thing... 
Imagine a member who has access only to service 1, I think this member should not be able to create service subscriptions using plan_ids that belong to service 2 (to which he doesn't have access to). I haven't tested this scenario, but it should be tested and verified IMO.

nidhi-soni1104 commented:
@daria Test that attempting to subscribe to a service plan that doesn't belong to the buyer's provider should result in a failure (404).
  
  I have added this test also, permission thing is not needed
  
  test_failure_subscribe_to_external_service_plan
  
  
  Also I added test for update method as well

nidhi-soni1104 commented:
@mayorova what you think of this PR?

mayorova commented:
@nidhi-soni1104 I've seen the following issues:

JSON representation of the service contract contains just two fields:
![Screenshot from 2024-03-19 12-58-21](https://github.com/3scale/porta/assets/1270328/819d95e3-3cd9-41d6-9224-b43cda587227)

Compare with the contents of the service contract in the "list" call:
![image](https://github.com/3scale/porta/assets/1270328/77b54a34-9e51-4e0e-8b3c-18a4e7b66062)

mayorova commented:
In XML calls, there are some weird `<to_xml/>` entries, which are not present in the LIST call.

![image](https://github.com/3scale/porta/assets/1270328/d8951c61-b789-4b9b-ae54-79aec688e60a)

mayorova commented:
Also, there is a difference in the error messages.

1. When updating the subscription to the plan that doesn't belong to the same service, the error is: 
```
<?xml version="1.0" encoding="UTF-8"?>
<error>Service plan must belong to the same product</error>
```


2. But when trying to subscribe to a service (with a POST), the error looks like this:
```
<?xml version="1.0" encoding="UTF-8"?>
<errors>
    <error>already subscribed to this service</error>
</errors>
```

I think we need to check what is the common format for the responses, and adhere to it.

For JSON the errors are:
```
{
    "error": "Service plan must belong to the same product"
}
```

vs

```
{
    "errors": {
        "base": [
            "already subscribed to this service"
        ]
    }
}
```

All these need to be covered with a test, also.

nidhi-soni1104 commented:
@mayorova  JSON representation of the service contract , now it will show all attributes

![image](https://github.com/3scale/porta/assets/84672731/647c9ea0-ca9d-427a-a963-ad466f68d39a)
![image](https://github.com/3scale/porta/assets/84672731/3ae5ad01-fdf2-4c14-bd7e-d7cb3a8e2708)



I checked the reason behind the difference in the format of errors so 
for(Service plan must belong to the same product)  we are using render_error from the API itself and already subscribed to this service is error message from a validation **one_contract_per_service** which is defined from here (https://github.com/3scale/porta/blob/43cbb51cc982ffb39e36023d0eea3223cef6ea99/app/lib/logic/contracting.rb#L108) 

So if I will use render_error in one_contract_per_service I will not be able to use it as API is inherited by **Admin::Api::ServiceBaseController**
 and render_error in **class ApplicationController < ActionController::Base** 
 
I am not sure about best way to use this render_error because we have module name **ErrorHandling** also  What do you think could be the best way?

nidhi-soni1104 commented:
@mayorova what do you think about the above comment?

mayorova commented:
Superseded by https://github.com/3scale/porta/pull/3759
